Incorrect C code generated when subscripted input variable is not referenced elsewhere

If you have a model that specifies an input variable (in the model spec `inputVarNames`), and that variable has subscripts, AND that variable is not referenced (i.e., it would otherwise be pruned due to it not being in the dependency chain of an output variable), then SDE will generate C code that references that variable, but there won't be a correct declaration for the variable (because it is getting pruned prematurely), and the build will fail when compiling the C code.

This is a bit of an edge case, because in this scenario such a slider would have no impact on the model behavior, but it would still be good to correct the issue because it can be confusing for a user of SDE to encounter a compilation error (and there is no other guidance to explain why it is failing).

The fix is to change the `removeUnusedVariables` code to ignore the subscript part when processing `inputVars` just like we do for `outputVars`.  I will add an automated test case to cover this scenario.

Example `inputs.csv`:
```csv
2,slider,view1,example slider[cn],example slider,,Assumptions,1,10,5,1,years,,,,,,,,,,,,,,,,Description goes here
```

Error message:
```
Input file inputs.csv has been changed
Generating files...
Done generating files (0.0s)
Generating model...
    Generating WebAssembly module
    build/processed.c:13500:6: error: use of undeclared identifier '_example_slider'
     13500 |
    &_example_slider[2],
                 |            ^
    build/processed.c:13518:3: error: use of undeclared identifier '_example_slider'
     13518 |     _example_slider[2] = i
    nputData[1];
                 |     ^
    2 errors generated.
    emcc: error: '/Users/campbell/Projects/CI/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -Werror=implicit-function-declaration --sysroot=/Users/campbell/Projects/CI/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -Ibuild -Wall -Os build/processed.c -c -o /var/folders/8r/0kjp9h6s649gf7k_p66rd7mr0000gn/T/emscripten_temp_43ba8k3p/processed_0.o' failed (returned 1)

ERROR: Failed to compile C model to WebAssembly: emcc command failed (code=1)
        at buildWasm (file:///Users/campbell/Projects/CI/example/node_modules/@sdeverywhere/plugin-wasm/dist/index.js:99:11)
        at async WasmPlugin.postGenerateC (file:///Users/campbell/Projects/CI/example/node_modules/@sdeverywhere/plugin-wasm/dist/index.js:24:5)
        at async generateModel (file:///Users/campbell/Projects/CI/example/node_modules/@sdeverywhere/build/dist/index.js:518:18) 
```

--- a/packages/compile/src/generate/gen-code-c.spec.ts
+++ b/packages/compile/src/generate/gen-code-c.spec.ts
@@ -63,7 +63,68 @@ describe('generateCode (Vensim -> C)', () => {
       y = :NOT: x ~~|
     `
     const code = readInlineModelAndGenerateC(mdl)
-    expect(code).toMatch('#include "sde.h"')
+    expect(code).toMatch(`\
+#include "sde.h"
+
+// Model variables
+double _x;
+double _y;`)
+  })
+
+  it('should work when valid input variable name without subscript is provided in spec file', () => {
+    const mdl = `
+      x = 10 ~~|
+      y = x + 1 ~~|
+    `
+    const code = readInlineModelAndGenerateC(mdl, {
+      inputVarNames: ['x'],
+      outputVarNames: ['y']
+    })
+    expect(code).toMatch(`\
+#include "sde.h"
+
+// Model variables
+double _x;
+double _y;`)
+  })
+
+  it('should work when valid input variable name with subscript (referenced by output variable) is provided in spec file', () => {
+    const mdl = `
+      DimA: A1, A2 ~~|
+      A[DimA] = 10, 20 ~~|
+      B[DimA] = A[DimA] + 1 ~~|
+    `
+    const code = readInlineModelAndGenerateC(mdl, {
+      inputVarNames: ['A[A1]'],
+      outputVarNames: ['B[A1]', 'B[A2]']
+    })
+    expect(code).toMatch(`\
+#include "sde.h"
+
+// Model variables
+double _a[2];
+double _b[2];`)
+  })
+
+  it('should work when valid input variable name with subscript (not referenced by output variable) is provided in spec file', () => {
+    // Note that `A` is specified as an input variable, but `A` is not referenced by output
+    // variable `B`, which is an unusual (but valid) usage scenario, so `A` should not be
+    // pruned by the `removeUnusedVariables` code (see #438)
+    const mdl = `
+      DimA: A1, A2 ~~|
+      A[DimA] = 10, 20 ~~|
+      B[DimA] = 30, 40 ~~|
+    `
+    const code = readInlineModelAndGenerateC(mdl, {
+      inputVarNames: ['A[A1]'],
+      outputVarNames: ['B[A1]', 'B[A2]']
+    })
+    expect(code).toMatch(`\
+#include "sde.h"
+
+// Model variables
+double _a[2];
+double _b[2];`)
   })
 
   it('should throw error when unknown input variable name is provided in spec file', () => {

--- a/packages/compile/src/generate/gen-equation-c.spec.ts
+++ b/packages/compile/src/generate/gen-equation-c.spec.ts
@@ -743,6 +743,24 @@ describe('generateEquation (Vensim -> C)', () => {
     expect(genC(vars.get('_b'), 'init-constants')).toEqual(['_b = 30.0;'])
   })
 
+  it('should work when valid input variable name with subscript is provided in spec file', () => {
+    const vars = readInlineModel(
+      `
+        DimA: A1, A2 ~~|
+        A[DimA] = 10, 20 ~~|
+        B[DimA] = A[DimA] + 1 ~~|
+      `,
+      {
+        inputVarNames: ['A[A1]'],
+        outputVarNames: ['B[A1]', 'B[A2]']
+      }
+    )
+    expect(vars.size).toBe(3)
+    expect(genC(vars.get('_a[_a1]'), 'init-constants')).toEqual(['_a[0] = 10.0;'])
+    expect(genC(vars.get('_a[_a2]'), 'init-constants')).toEqual(['_a[1] = 20.0;'])
+    expect(genC(vars.get('_b'), 'eval')).toEqual(['for (size_t i = 0; i < 2; i++) {', '_b[i] = _a[i] + 1.0;', '}'])
+  })
+
   it('should throw error when unknown input variable name is provided in spec file', () => {
     expect(() =>
       readInlineModel(

--- a/packages/compile/src/model/model.js
+++ b/packages/compile/src/model/model.js
@@ -468,7 +468,11 @@ function removeUnusedVariables(spec) {
 
   // Keep all input variables
   for (const inputVarName of spec.inputVars) {
-    for (const v of varsWithName(inputVarName)) {
+    // The inputVars can include a raw index, e.g. `_input_var[0]`,
+    // which isn't an actual "ref id", so we'll just derive the
+    // var name by chopping off the index part.
+    const inputVarBaseName = inputVarName.split('[')[0]
+    for (const v of varsWithName(inputVarBaseName)) {
       recordUsedVariable(v)
     }
   }

